### ISSUE TITLE
Attention Visualisation

### ISSUE THREAD

**Issue by bikashg** [spro/practical-pytorch#31]
The diagrams present for attention visualisation don't show the network has learned very well. For eg : in the first figure, the output word "she" has the most attention to the input word "cinq", which is not correct. And in general, the learning seems to be very bad based on the attention visualisations. On the other hand, the training graph shows that the loss is very low (around 0.6); so the network must have learnt well. 

I am confused by these conflicting understanding. Could you please help explain it.

**Comment by spro** [comment]
The RNN is gradually building up its state, not just based on the current word, so as it gets further in the sentence it may become "more sure" about the context of previous words. That's why you sometimes see this shift. In some cases you will see the final EOS token being overly focused on because that final state encodes all the information in the sentence (as in seq2seq without attention).

Plus, the examples shown were only trained for around 40 minutes with a limited dataset, so the results are not great. With a small training set the network sometimes learns weird patterns that don't seem to line up at all, because it doesn't have to rely on alignment. Longer training time, more data, and a bidirectional encoder (coming soon) will improve results.

The bidirectional encoder especially helps because the forward and backward RNN states are combined to incorporate info from both directions. This makes the position of the attention more interpretable:

![screen shot 2017-05-30 at 7 41 17 pm](https://cloud.githubusercontent.com/assets/211613/26756671/f112a5d2-485c-11e7-93d2-56d72cf63d71.png)
